src/main/src/main/java/org/geoserver/security/auth/EnvPwdResolverUserGroupServiceWrapper.java

```diff
@@ -0,0 +1,95 @@
+/* (c) 2022 Open Source Geospatial Foundation - all rights reserved
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.security.auth;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.apache.commons.lang.StringUtils;
+import org.geoserver.platform.GeoServerEnvironment;
+import org.geoserver.platform.GeoServerExtensions;
+import org.geoserver.security.GeoServerSecurityManager;
+import org.geoserver.security.GeoServerUserGroupService;
+import org.geoserver.security.impl.GeoServerUser;
+import org.geoserver.security.password.GeoServerPasswordEncoder;
+import org.geotools.util.logging.Logging;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+
+/**
+ * A UserGroupService wrapper able to resolve password that have been parametrized. The placeholder
+ * can be encoded in plain txt or any reversible encoding.
+ */
+class EnvPwdResolverUserGroupServiceWrapper implements UserDetailsService {
+
+    private GeoServerUserGroupService delegate;
+    private final GeoServerEnvironment environment;
+
+    private static Logger LOGGER = Logging.getLogger(EnvPwdResolverUserGroupServiceWrapper.class);
+
+    EnvPwdResolverUserGroupServiceWrapper(GeoServerUserGroupService userDetailsService) {
+        this.delegate = userDetailsService;
+        environment = GeoServerExtensions.bean(GeoServerEnvironment.class);
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        UserDetails userDetails = delegate.loadUserByUsername(username);
+        if (environment != null && GeoServerEnvironment.allowEnvParametrization()) {
+            if (userDetails instanceof GeoServerUser) {
+                GeoServerUser user = (GeoServerUser) userDetails;
+                String password = user.getPassword();
+                GeoServerSecurityManager manager = delegate.getSecurityManager();
+                List<GeoServerPasswordEncoder> encoders =
```

Comment:
I'm not familiar with how password encoding works, but it seems you're getting a UserDetails with an encoded password, you decode it and replace it with a clear text version. But the upper levels should still suppose the password is encoded, or not? Also, I've seen in the past that plain passwords are normally prefixed by "plain:"....

Reply:
The encoded password (including plain) is first decoded to retrieve the eventually present placeholder. Then the placeholder is resolved on env. The env value provided might be encoded in wathever encryption supported by GeoServer since the dao authentication provider will try all the password encoders to match the incoming password with the resolved one. The only component that will have visibility over the resolved value is the DaoAuthenticationProvider wrapped into the UserNamePasswordAtuehtnicationProvider, that will actually do the password check.
I've also added an example of passwords that might be stored in the properties on documentation.